packages/firestore/src/model/proto_values.ts

```diff
@@ -0,0 +1,385 @@
+/**
+ * @license
+ * Copyright 202 Google LLC
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import * as api from '../protos/firestore_proto_api';
+
+import { TypeOrder } from './field_value';
+import { assert, fail } from '../util/assert';
+import { keys, size } from '../util/obj';
+import { ByteString } from '../util/byte_string';
+import {
+  numericComparator,
+  numericEquals,
+  primitiveComparator
+} from '../util/misc';
+
+// A RegExp matching ISO 8601 UTC timestamps with optional fraction.
+const ISO_TIMESTAMP_REG_EXP = new RegExp(
+  /^\d{4}-\d\d-\d\dT\d\d:\d\d:\d\d(?:\.(\d+))?Z$/
+);
+
+/** Extracts the backend's type order for the provided value. */
+export function typeOrder(value: api.Value): TypeOrder {
+  if ('nullValue' in value) {
```

Comment:
It seems like we could take the first key in `value` and then look this up in a map. Something like:

```ts
const typeOrderMap = {
  'nullValue': TypeOrder.NullValue,
  // ...
}

export function typeOrder(value: api.Value): TypeOrder {
  for (const key in value) {
    const result = typeOrderMap[key];
    if (result !== undefined) {
      return result;
    }
  }
  return fail('Invalid value type: ' + JSON.stringify(value));
}
```